firmware/quintuna/CRIT/src/app/app_leds.c

```diff
@@ -0,0 +1,177 @@
+
+static uint32_t boardStatusToRGBBits(BoardLEDStatus status, uint32_t redMask, uint32_t greenMask, uint32_t blueMask)
+{
+    switch (status)
+    {
+        case BOARD_LED_STATUS_OK:
+            return greenMask;
+
+        case BOARD_LED_STATUS_WARNING:
+            // Yello = red + green
+            return (redMask | greenMask);
+
+        case BOARD_LED_STATUS_FAULT:
+            return redMask;
+
+        case BOARD_LED_STATUS_MISSING_HEARTBEAT:
+            return blueMask;
+
+        default:
+            return 0;
+    }
+}
+
+static BoardLEDStatus worstBoardStatus(CanAlertBoard board)
+{
+    bool is_missing_heartbeat = false;
+
+    switch (board)
+    {
+        case BMS_ALERT_BOARD:
+            is_missing_heartbeat = app_canAlerts_CRIT_Fault_MissingBMSHeartbeat_get();
+            break;
+        case FSM_ALERT_BOARD:
+            is_missing_heartbeat = app_canAlerts_CRIT_Fault_MissingFSMHeartbeat_get();
+            break;
+        case RSM_ALERT_BOARD:
+            is_missing_heartbeat = app_canAlerts_CRIT_Fault_MissingRSMHeartbeat_get();
+            break;
+        case VC_ALERT_BOARD:
+            is_missing_heartbeat = app_canAlerts_CRIT_Fault_MissingVCHeartbeat_get();
+            break;
+        // TODO: add DAM
+        default:
+            is_missing_heartbeat = false;
+            break;
+    }
+
+    if (is_missing_heartbeat)
+    {
+        return BOARD_LED_STATUS_MISSING_HEARTBEAT;
+    }
+    else if (app_canAlerts_BoardHasFault(board))
+    {
+        return BOARD_LED_STATUS_FAULT;
+    }
+    else if (app_canAlerts_BoardHasWarning(board))
+    {
+        return BOARD_LED_STATUS_WARNING;
+    }
+
+    return BOARD_LED_STATUS_OK;
+}
+
+// -----------------------------------------------------------------------------
+// 4) app_leds_update
+//    The main function that assembles all LED bits and shifts them out.
+// -----------------------------------------------------------------------------
+void app_leds_update(void)
+{
+    const bool imd_fault_latched  = app_canRx_BMS_ImdLatchedFault_get();
+    const bool bspd_fault_latched = app_canRx_BMS_BspdLatchedFault_get();
+    const bool ams_fault_latched  = app_canRx_BMS_BmsLatchedFault_get();
+    const bool push_drive         = (app_canRx_VC_State_get() == VC_DRIVE_STATE);
+    const bool regen_light_on     = app_canRx_VC_RegenEnabled_get();
+    const bool torquevec_light_on = app_canRx_VC_TorqueVectoringEnabled_get();
+    const bool is_shdn_ok         = (app_canRx_VC_FirstFaultNode_get() == SHDN_OK);
+
+    // Determine board status.
+    const BoardLEDStatus bms_status  = worstBoardStatus(BMS_ALERT_BOARD);
+    const BoardLEDStatus fsm_status  = worstBoardStatus(FSM_ALERT_BOARD);
+    const BoardLEDStatus rsm_status  = worstBoardStatus(RSM_ALERT_BOARD);
+    const BoardLEDStatus vc_status   = worstBoardStatus(VC_ALERT_BOARD);
+    const BoardLEDStatus crit_status = worstBoardStatus(CRIT_ALERT_BOARD);
+    // TODO: add DAM
+
+    uint32_t ledMask = 0;
```

Comment:
this is one way to approach it. Is it possible that you could also consider using a struct with bit fields? I think that makes this much more readable